test/test_images/latency/main.go

```diff
@@ -0,0 +1,249 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/transport/http"
+	"github.com/knative/eventing/test/performance"
+)
+
+const (
+	defaultEventType   = "perf-test-event-type"
+	defaultEventSource = "perf-test-event-source"
+
+	// The interval and timeout used for polling in checking event states.
+	pollInterval = 1 * time.Second
+	pollTimeout  = 4 * time.Minute
+)
+
+// flags for the image
+var (
+	sinkURL            string
+	msgSize            int
+	eventNum           int
+	errorRateThreshold float64
+	timeout            int
+	encoding           string
+	eventMap           sync.Map
+)
+
+// eventStatus is status of the event, for now only if all events are in received status can the
+// test be considered as PASS.
+type eventStatus int
+
+const (
+	sent eventStatus = iota
+	received
+	undelivered
+	dropped
+	duplicated
+	corrupted // TODO(Fredy-Z): corrupted status is not being used now
+)
+
+// state saves the data that is used to generate the metrics
+type state struct {
+	sentTime int64
+	latency  int64
+	status   eventStatus
+}
+
+func init() {
+	flag.StringVar(&sinkURL, "sink", "", "The sink URL for the event destination.")
+	flag.IntVar(&msgSize, "msg-size", 100, "The size of each message we want to send. Generate random strings to avoid caching.")
+	flag.IntVar(&eventNum, "event-num", 10, "The number of events we want to send.")
+	flag.IntVar(&timeout, "timeout", 30, "Timeout in seconds. If we do not receive a message back within a limited time, we consider it to be dropped.")
+	flag.Float64Var(&errorRateThreshold, "error-rate-threshold", 0.1, "Rate of error event deliveries we allow. We fail the test if the error rate crosses the threshold.")
+	flag.StringVar(&encoding, "encoding", "binary", "The encoding of the cloud event, one of(binary, structured).")
+}
+
+func receivedEvent(event cloudevents.Event) {
+	eventID := event.ID()
+	if eventState, ok := eventMap.Load(eventID); ok {
+		state := eventState.(*state)
+		switch state.status {
+		// if the status is sent, calculate the latency
+		case sent:
+			now := time.Now().UTC().UnixNano()
+			state.latency = now - state.sentTime
+			state.status = received
+		// if the status is received or duplicated, consider the event to be duplicate
+		case received, duplicated:
+			state.status = duplicated
```

Comment:
Setting `state.status` here isn't thread-safe: it updates concurrently with `waitForEventStatesReady`.